### PR TITLE
Refactor build scripts and npm package process

### PR DESCRIPTION
**Pull request by etrepum**
Addresses refactoring of npm/build processes per #5869 

- [x] Adds some tests to sanity check packages (their package.json, the existence of flow and docs)
- [x] Refactors basically all of scripts/* to use a generated model for the available packages
- [x] Update scripts/www/rewriteImports.js to use the new infrastructure
- [x] Generate tsconfig.json paths (npm run update-tsconfig)
- [x] Generate .flowconfig paths (npm run update-flowconfig)
- [x] Generate boilerplate `README.md` and `docs/packages/{package}.md` files (npm run update-docs)
- [x] jest config
- [x] viteModuleResolution
- [x] docusaurus config
- [x] Started a [Maintainers' Guide](https://lexical-mzg4hwe1m-fbopensource.vercel.app/docs/maintainers-guide) section in the docs (currently fairly hidden, only referenced in the repo README)
- [x] Adds build integration tests to make sure all of the examples and several other framework scenarios (astro, svelte, next.js) build and run with release artifacts

RE #5869 this mostly addresses the "Reduce number of changes needed to add new package" - since I have not added a new package myself yet I think another maintainer (@StyleT?) could add those docs to the guide using this as a baseline?

I was also thinking about maybe starting on an eslint plugin in the next week or two to help with the "rules of $functions" so if someone else doesn't get to it first I could add that documentation then when I am more familiar with all of those details.

Closes #5940 (parts of this PR were extracted there to maybe expedite review)

### CODE REVIEW

**Comment by etrepum** [comment]
@StyleT I wanted to put this up earlier on in case you had any feedback on this approach

File: scripts/www/rewriteImports.js [hunk 1]
```diff
@@ -8,72 +8,78 @@
 
 'use strict';
 
+const fs = require('fs-extra');
 const glob = require('glob');
-const {readFile, writeFile} = require('fs');
+const path = require('node:path');
+const {packagesManager} = require('../shared/packagesManager');
+const npmToWwwName = require('./npmToWwwName');
+const {t, transform} = require('hermes-transform');
```
Comment (etrepum) [inline scripts/www/rewriteImports.js:16]: Oh boy it was a journey to find a way to correctly transform modern flow files. ast-types is not up to date with flow's features so jscodeshift/recast can't print transformed code. I was lucky to find this used by react-native.

File: scripts/www/rewriteImports.js [hunk 2]
```diff
@@ -8,72 +8,78 @@
...
+          context.replaceNode(node.source, t.StringLiteral({value}));
+        }
+      },
+      Program(node) {
+        if (
+          node.docblock &&
+          node.docblock.comment &&
+          node.docblock.comment.value.includes('@flow strict')
+        ) {
+          node.docblock.comment.value = node.docblock.comment.value.replace(
```
Comment (etrepum) [inline scripts/www/rewriteImports.js:54]: I don't think it's correct to mutate this in-place, but I couldn't quickly figure out a mutator incantation that didn't fail… and this produced the correct output. Can't be worse than some regexes right? 😆 

Maybe someone inside meta could refactor this to do it the right way.

File: tsconfig.json [hunk 3]
```diff
@@ -12,212 +12,207 @@
...
+      "@lexical/text/src": ["./packages/lexical-text/src"],
+      "@lexical/utils/src": ["./packages/lexical-utils/src"],
+      "@lexical/yjs/src": ["./packages/lexical-yjs/src"],
+      "shared/src": ["./packages/shared/src"]
     }
   },
   "include": ["./libdefs", "./packages"],
   "exclude": [
+    "./libdefs/*.js",
     "**/lexical-*/src/__tests__/**",
```
Comment (etrepum) [inline tsconfig.json:210]: I'd like to turn these on but the tests have all sorts of broken types and fixing them would be some big scope creep!

File: packages/lexical-website/docs/maintainers-guide.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+## Package naming conventions
+
+### Overall
+
+| Usage | Convention |
+| -- | -- |
+| Directory name | `packages/lexical-package-name` |
+| Entrypoint | `packages/lexical-package-name/index.ts` |
+| Flow types | `packages/lexical-package/flow/LexicalPackage.js.flow` |
```
Comment (etrepum) [inline packages/lexical-website/docs/maintainers-guide.md:44]: Does anyone happen to know if those `packages/lexical-package/LexicalPackage.js` files are used for anything? Is that for www tooling? They would be easy to generate but I think it would be even better to get rid of them :)

They are the files that boil down to `module.exports = require('./dist/LexicalPackage.js');`

**Review by StyleT — COMMENTED** [review]
Hi! And thanks for this work!

1. Do you think we can split this PR onto a few smaller ones? I see changes here that are seemingly unrelated (like creation of the maintainers' guide). Dunno how we can merge it all at one w/o taking a leap of faith ;P 
2. I would prefer to keep all that is related to the Lexical contribution within `[CONTRIBUTING.md](https://github.com/facebook/lexical/blob/main/CONTRIBUTING.md)` Unless we have a strong reason to do it differently
3. This PR seemingly misses TODO steps on how to add a new package

**Comment by etrepum** [comment]
@StyleT

1. In a way, the maintainers guide is documentation about the tools that this PR either updates or adds. I could add that file separately, but I'm not sure what the point would be. I could certainly split up the work into multiple commits, e.g. "all of the tools" and "automated modifications to configuration" could be separate, but otherwise this is all rather intertwined.
2. The primary reason I chose to put it in lexical-website was that the CONTRIBUTING.md isn't rendered anywhere on lexical.dev, and the intended audience for this document are people working a lot closer to the build infrastructure than almost all contributors would need to worry about. It seemed like a better idea to keep CONTRIBUTING.md short so it's not intimidating.
3. Per creating a new module, that I would either like someone else to work on, or I will handle that separately in the course of creating a new module. I haven't added a new module yet, so I am not confident that I have all of the required steps in order, especially things that are outside of the source tree like the PR template and other meta concerns. There is some commentary about that in the description of this PR.

File: scripts/__tests__/unit/build.test.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+          ),
+        ).toEqual([]);
+      });
+      it('monorepo dependencies must use the exact monorepo version', () => {
+        Object.entries(dependencies).forEach(([dep, version]) => {
+          if (publicNpmNames.has(dep)) {
+            expect([dep, version]).toEqual([dep, monorepoPackageJson.version]);
+          }
+        });
+      });
```
Comment (etrepum) [inline scripts/__tests__/unit/build.test.js:66]: This is a test that would've caught the version issue of #5901

File: scripts/npm/prepare-release.js [hunk 6]
```diff
@@ -10,46 +10,48 @@
...
-    await exec(`cp -R LICENSE ./packages/${pkg}/npm`);
-    await exec(`cp -R ./packages/${pkg}/README.md ./packages/${pkg}/npm`);
-  }
+/**
+ * Prepares the npm directories for all public packages. If a package should
+ * not be published, make sure to set `"private": true` in its package.json.
+ */
+function preparePublicPackages() {
+  packagesManager.getPublicPackages().forEach(preparePackage);
 }
```
Comment (etrepum) [inline scripts/npm/prepare-release.js:55]: This is the part that would've published @lexical/devtools-core (re #5901)

**Comment by etrepum** [comment]
happy to rebase and fix conflicts on this one when someone gives me a heads-up (on here or discord) that they're ready to review and consider for merge, I don't want to have to do it too many times

File: packages/lexical-website/docs/maintainers-guide.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+| -- | -- |
+| Directory name | `packages/lexical-package-name` |
+| Entrypoint | `packages/lexical-package-name/src/index.{ts,tsx}` |
+| Flow types | `packages/lexical-package/flow/LexicalPackageName.js.flow` |
+| package.json name | `@lexical/package-name` |
+| Documentation | `packages/lexical-package-name/README.md` |
+| Unit Tests | `packages/lexical-package-name/src/__tests__/unit/LexicalPackageName.test.{ts,tsx}` |
+| dist (gitignore'd build product) | `packages/lexical-package-name/dist` |
+| npm (gitignore'd prerelease product) | `packages/lexical-package-name/npm` |
+| www entrypoint? | `packages/lexical-package-name/LexicalPackageName.js` |
```
Comment (acywatson) [inline packages/lexical-website/docs/maintainers-guide.md:50]: yes

File: scripts/clean.js [hunk 8]
```diff
@@ -9,12 +9,13 @@
...
-  fs.removeSync(path.resolve(`./.ts-temp`));
-  fs.removeSync(path.resolve(`./packages/${pkg}/dist`));
-  fs.removeSync(path.resolve(`./packages/${pkg}/npm`));
-});
+fs.removeSync(path.resolve(`./npm`));
+fs.removeSync(path.resolve(`./.ts-temp`));
+packagesManager
+  .getPublicPackages()
+  .forEach((pkg) =>
+    ['dist', 'npm'].forEach((subdir) => fs.removeSync(pkg.resolve(subdir))),
```
Comment (acywatson) [inline scripts/clean.js:20]: don't we cover npm on line 15?

**Comment by acywatson** [comment]
This looks good to me - huge improvement.

I honestly don't know if all the WWW and flow stuff is right, but I think the fastest way to tell is to merge and fix any issues that come up in the sync.

I'll run a sync internally ASAP, then we can merge this and do a sync with just this PR so we can catch any issues.

**Review by acywatson — APPROVED** [review]

File: packages/lexical-website/docs/maintainers-guide.md (hunk 7 above)
Comment (etrepum) [reply packages/lexical-website/docs/maintainers-guide.md]: great, just added a test to check for them and a script to create the stubs

File: scripts/clean.js (hunk 8 above)
Comment (etrepum) [reply scripts/clean.js:20]: no, line 15 covers the monorepo's npm. This is something I added in order to facilitate the build integration tests. It hosts the `npm pack` output from each of the public packages

```
❯ ls npm
lexical-0.14.5.tgz               lexical-hashtag-0.14.5.tgz       lexical-mark-0.14.5.tgz          lexical-rich-text-0.14.5.tgz
lexical-clipboard-0.14.5.tgz     lexical-headless-0.14.5.tgz      lexical-markdown-0.14.5.tgz      lexical-selection-0.14.5.tgz
lexical-code-0.14.5.tgz          lexical-history-0.14.5.tgz       lexical-offset-0.14.5.tgz        lexical-table-0.14.5.tgz
lexical-devtools-core-0.14.5.tgz lexical-html-0.14.5.tgz          lexical-overflow-0.14.5.tgz      lexical-text-0.14.5.tgz
lexical-dragon-0.14.5.tgz        lexical-link-0.14.5.tgz          lexical-plain-text-0.14.5.tgz    lexical-utils-0.14.5.tgz
lexical-file-0.14.5.tgz          lexical-list-0.14.5.tgz          lexical-react-0.14.5.tgz         lexical-yjs-0.14.5.tgz
```

File: packages/lexical-website/docs/maintainers-guide.md (hunk 7 above)
Comment (acywatson) [reply packages/lexical-website/docs/maintainers-guide.md]: Oh yea but do you need the question mark in there now that you know lol

Comment (etrepum) [reply packages/lexical-website/docs/maintainers-guide.md]: it got removed but github always shows comments in the context of the commit they were written about https://github.com/facebook/lexical/pull/5876/commits/1e2a4960c330fd14491f2f09756023c3f15dfeb5#diff-d575e613cfb9cf9256293e9cb0f838e5b4ddcf3890f11aed370bbc03f15cd3bcR50

File: scripts/update-docs.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    # \`${npmName}\`
+
+[![See API Documentation](https://lexical.dev/img/see-api-documentation.svg)](https://lexical.dev/docs/api/modules/${apiModuleName})
+
+${description}
+`.trim() + '\n'
+  );
+}
+
+function updateDocs() {
```
Comment (StyleT) [inline scripts/update-docs.js:43]: So as far as I understand this script doesn't update anything but rather creates files only. I see the following issues here:

1. It can't keep existing docs consistent
2. We run it at a wrong time. `update-packages` is a bit too late as "invalid" package already got into `main` branch. And it wouldn't help package creator to see that this script shall be run (if I understand it correctly)

I think it's better to opt for Docusaurus plugin and generate it on the fly (same approach we use for API docs). Plus extend `scripts/__tests__/unit/build.test.js` to verify that all public packages have README

Here is a crude example of how this can be done: https://github.com/facebook/lexical/compare/main...feature/docs_packages_autogen
It lacks watch mode + ability to set package order + custom sidebar name w/o affecting original README.md
It was inspired by https://github.com/tgreyuk/typedoc-plugin-markdown/blob/master/packages/docusaurus-plugin-typedoc/src/plugin.ts#L18

Comment (StyleT) [reply scripts/update-docs.js:43]: btw if we keep it as is - may be better to rename it to "create-docs"

File: scripts/update-tsconfig.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    filepath: jsonFileName,
+  });
+  if (prevTsconfigContents !== nextTsconfigContents) {
+    fs.writeFileSync(jsonFileName, nextTsconfigContents);
+  }
+}
+
+async function updateAllTsconfig() {
+  const prettierConfig = (await prettier.resolveConfig('./')) || {};
+  await updateTsconfig({
```
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/update-tsconfig.js:90]: That's fine, we can add more files if that's the preferred style. I was just following the existing convention.

File: scripts/__tests__/integration/prepare-release.test.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        'npm',
+        `${pkg.getDirectoryName()}-${pkg.packageJson.version}.tgz`,
+      );
+      test(`${tgzPath} exists`, () =>
+        expect(fs.existsSync(tgzPath)).toBe(true));
+    });
+  }
+});
+['examples', 'scripts/__tests__/integration/fixtures']
+  .flatMap((packagesDir) => glob.sync(`${packagesDir}/*/package.json`))
```
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/prepare-release.test.js:30]: They are real application packages, but not published to npm. Adding them as zip files would be something that could be done but then it is even harder to see why and how the tests broke. We can add them to examples but I did not do that here because I didn't want to have the discussions about whether these are officially endorsed or not since there are third party packages for some of these frameworks that offer better integration.

File: scripts/__tests__/integration/fixtures/lexical-esm-nextjs/package.json [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "dev": "next dev",
+    "build": "next build",
+    "start": "next start",
+    "test": "playwright test"
+  },
+  "dependencies": {
+    "@lexical/plain-text": "^0.14.5",
```
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/fixtures/lexical-esm-nextjs/package.json:12]: I did not take that approach because the same test code is used to verify that the examples work with build artifacts, which are necessarily projects that can be extracted from the monorepo and run elsewhere.

File: .prettierignore [hunk 13]
```diff
@@ -1,14 +1,16 @@
...
 
-packages/**/dist/**
-packages/**/build/**
-packages/**/npm/**
-packages/**/config/**
+**/dist/**
+**/build/**
+**/npm/**
+**/__tests__/integration/fixtures/**
+**/config/**
```
[the remark this answers is not in the document]
Comment (etrepum) [reply .prettierignore:7]: I have no opinion on config folders, this was just keeping the ignore files in sync between eslint and prettier

File: scripts/update-docs.js (hunk 9 above)
Comment (etrepum) [reply scripts/update-docs.js:43]: update-packages is supposed to be run before anything is merged to main, it is not too late, when creating a new project you would also need it to generate other configuration. If you see the WIP eslint PR it has some documentation about this https://github.com/facebook/lexical/blob/690f73920a9314805bef9f9ceb096e3eda5f4e00/packages/lexical-website/docs/maintainers-guide.md#creating-a-new-package

We could add a docs autogen plugin but I was just going with the existing conventions.

Either way, watch doesn't work when editing those README files because the markdown transclusion doesn't support watch of the transcluded files (this predated the work that I have done).

Comment (etrepum) [reply scripts/update-docs.js:43]: Oh and there are tests that will fail if you try and add a new package without somehow creating all of the required boilerplate files, so it would be hard to land something on main before running update-packages or equivalent.

File: .prettierignore (hunk 13 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply .prettierignore]: Removed from both in 1614a6047575ce7a980ace2ca8cc8209bfcd7165 since there are no config folders to begin with

File: scripts/update-tsconfig.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/update-tsconfig.js:93]: Extracted to tsconfig*.paths.json 224e201808b701337bad2a2dedfec4b4ca645be3

File: scripts/__tests__/integration/prepare-release.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (StyleT) [reply scripts/__tests__/integration/prepare-release.test.js:30]: > They are real application packages, but not published to npm. Adding them as zip files would be something that could be done but then it is even harder to see why and how the tests broke. We can add them to examples but I did not do that here because I didn't want to have the discussions about whether these are officially endorsed or not since there are third party packages for some of these frameworks that offer better integration.

Makes sense. Then can we give them more of a semantic naming so it's clear what exactly we test with every one of them. I'm puzzled why exactly svelte/astro/etc is important here.

File: scripts/update-tsconfig.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (StyleT) [reply scripts/update-tsconfig.js:93]: > 224e201

Awesome! I would still suggest to name it `tsconfig*.generated.json` to emphasize that they are not manually written

[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/update-tsconfig.js:93]: Had to do a follow-up fix since the jest config expected a specific layout of tsconfig.json 262fed01f88c4d7e3bbc525c115016f9eea259b2

File: scripts/update-docs.js (hunk 9 above)
Comment (etrepum) [reply scripts/update-docs.js]: renamed update-docs and update-www-stubs to create-docs and create-www-stubs in ddf3d3fb73561a408dde6e9dd9fdcbc78abee739

Comment (StyleT) [reply scripts/update-docs.js]: > Either way, watch doesn't work when editing those README files because the markdown transclusion doesn't support watch of the transcluded files (this predated the work that I have done).

So if you use approach similar to https://github.com/facebook/lexical/compare/main...feature/docs_packages_autogen with watch implemented in a way similar to https://github.com/tgreyuk/typedoc-plugin-markdown/blob/master/packages/docusaurus-plugin-typedoc/src/plugin.ts#L97 - it will work. As then you don't need markdown transclusion.

Btw this lack of watch support caused me to lose few hours on debugging few weeks ago. But I wasn't paying attention to it.

P.S.: I'm ok to do it as follow up for sure

File: scripts/__tests__/integration/prepare-release.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (acywatson) [reply scripts/__tests__/integration/prepare-release.test.js:30]: > I'm puzzled why exactly svelte/astro/etc is important here.

Tons of people use these frameworks in open source

File: scripts/update-tsconfig.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/update-tsconfig.js:93]: it seems that esbuild does not support having an array for the extends argument so we can't have the build json extend both the main tsconfig and a generated file, so I will roll back those changes and someone can revisit splitting up the generated tsconfig separately.

```
[vite:esbuild] The "path" argument must be of type string. Received an instance of Array
file: /Users/bob/src/lexical/packages/lexical-playground/src/index.tsx
error during build:
TypeError: The "path" argument must be of type string. Received an instance of Array
    at validateString (node:internal/validators:162:11)
    at Object.isAbsolute (node:path:1160:5)
    at resolveExtends (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19612:22)
    at parseExtends (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19593:34)
    at parse$f (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19544:24)
    at async loadTsconfigJsonForFile (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19933:24)
    at async transformWithEsbuild (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19736:36)
    at async Object.transform (/Users/bob/src/lexical/node_modules/vite/dist/node/chunks/dep-6e2fe41e.js:19821:32)
    at async transform (/Users/bob/src/lexical/node_modules/rollup/dist/shared/rollup.js:22085:16)
    at async ModuleLoader.addModuleSource (/Users/bob/src/lexical/node_modules/rollup/dist/shared/rollup.js:22311:30)
```

File: scripts/__tests__/integration/fixtures/lexical-esm-nextjs/package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/fixtures/lexical-esm-nextjs/package.json:12]: The integration tests do verify that the a correct version of lexical was installed, and updating the package.json or package-lock.json are not really necessary in this scenario (unless updating the other dependencies, e.g. the frameworks) since we are essentially linking our versions of the monorepo package artifacts into node_modules.

https://github.com/facebook/lexical/pull/5876/files#diff-19bab06b607c228fd4b4b2a97b16cbc72e6ad4b656771cdf497172152339ee64R124-R140

File: scripts/update-docs.js (hunk 9 above)
Comment (etrepum) [reply scripts/update-docs.js]: I also lost some time to the lack of watch support for README.md files and resolving it is something I would work on if you don't get to it first :)

File: scripts/__tests__/integration/prepare-release.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/prepare-release.test.js:30]: I would be open to cleaning up the test fixtures for these frameworks and making them "full" examples, but the goal of adding them here was to prevent future compatibility regressions since I've caused and fixed so many along the way to getting esm support. It is really hard to test so many combinations of frameworks that all have their own special way of using webpack/swc/esbuild/vite/etc. unless it's already in the monorepo test suite.

Angular would be a good one to do as a follow-up. The choice of next.js, sveltekit and astro were made because we are not testing them with examples in-tree (we are only testing a basic vite config with and without react), and we have gotten issues or discord problem reports about those frameworks specifically. I had them sitting around for my own testing.

[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/prepare-release.test.js:30]: What we are testing with them right now is simple:
* does lexical build with this framework
* did lexical render (tested by starting the build product & checking with playwright)

[the remark this answers is not in the document]
Comment (StyleT) [reply scripts/__tests__/integration/prepare-release.test.js:30]: Ok, got it. Sounds good to me.

Out of curiosity:
Can you give some example of any change to Lexical that won’t break vanilla js integration but will break any of these “non-React” frameworks

**Review by StyleT — APPROVED** [review]

File: scripts/__tests__/integration/prepare-release.test.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/prepare-release.test.js:30]: The regressions that I noticed were mostly around changes to the package.json files and the fork modules. Rather than vanilla js vs. non-react frameworks it's usually of a question of vite vs swr vs webpack plus whatever configuration to those bundlers come with the framework.

[the remark this answers is not in the document]
Comment (etrepum) [reply scripts/__tests__/integration/prepare-release.test.js:30]: #5823 is a concrete next.js issue that this test would've caught
